core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala

```diff
@@ -290,6 +290,10 @@ private[zio] final class FiberContext[E, A](
           opcount = opcount + 1
         }
       } catch {
+        case _: InterruptedException =>
+          Thread.interrupted
+          curIo = terminate(IO.interrupt)
```

Comment:
Should we make sure this IO runs on the async thread pool?

Reply:
That will automatically happen if it's supposed to. The `lock` uses ensuring to bump back to the previous thread pool after it's done. So "on the way out", when cleaning up after termination, if appropriate the fiber will jump to previous thread pools.